src/core/morphcontroller.cpp

```diff
@@ -0,0 +1,393 @@
+
+/*!
+    \qmlproperty weight7
+
+    Specifies the weight value at index 7.
+ */
+
+/*!
+    \qmlproperty morphWeights
+    \internal
+
+    Used by the animation aspect to set all weights at once.
+ */
+
+MorphController::MorphController(Qt3DCore::QNode *parent)
+    : Qt3DRender::QShaderData(parent)
+    , m_count(0)
+    , m_shouldEmitMorphWeights(true)
+{
+    std::fill(std::begin(m_weights), std::end(m_weights), 0.0f);
+
+    auto shouldEmitMorphWeightsChanged = [this]() {
+        if (m_shouldEmitMorphWeights)
+            emit morphWeightsChanged();
+    };
+
+    QObject::connect(this, &MorphController::weight0Changed,
+                     this, shouldEmitMorphWeightsChanged);
+    QObject::connect(this, &MorphController::weight1Changed,
+                     this, shouldEmitMorphWeightsChanged);
+    QObject::connect(this, &MorphController::weight2Changed,
+                     this, shouldEmitMorphWeightsChanged);
+    QObject::connect(this, &MorphController::weight3Changed,
+                     this, shouldEmitMorphWeightsChanged);
+    QObject::connect(this, &MorphController::weight4Changed,
+                     this, shouldEmitMorphWeightsChanged);
+    QObject::connect(this, &MorphController::weight5Changed,
+                     this, shouldEmitMorphWeightsChanged);
+    QObject::connect(this, &MorphController::weight6Changed,
+                     this, shouldEmitMorphWeightsChanged);
+    QObject::connect(this, &MorphController::weight7Changed,
+                     this, shouldEmitMorphWeightsChanged);
+}
+
+MorphController::~MorphController()
+{
+}
+
+float MorphController::weight0() const
+{
+    return m_weights[0];
+}
+
+float MorphController::weight1() const
+{
+    return m_weights[1];
+}
+
+float MorphController::weight2() const
+{
+    return m_weights[2];
+}
+
+float MorphController::weight3() const
+{
+    return m_weights[3];
+}
+
+float MorphController::weight4() const
+{
+    return m_weights[4];
+}
+
+float MorphController::weight5() const
+{
+    return m_weights[5];
+}
+
+float MorphController::weight6() const
+{
+    return m_weights[6];
+}
+
+float MorphController::weight7() const
+{
+    return m_weights[7];
+}
+
+int MorphController::count() const
+{
+    return m_count;
+}
+
+QVariantList MorphController::morphWeights() const
+{
+    using WeightGetterFunc = float (MorphController::*)() const;
+    const WeightGetterFunc weightGetterFuncs[]{
+        &MorphController::weight0,
+        &MorphController::weight1,
+        &MorphController::weight2,
+        &MorphController::weight3,
+        &MorphController::weight4,
+        &MorphController::weight5,
+        &MorphController::weight6,
+        &MorphController::weight7
+    };
+
+    QVariantList weights;
+    weights.reserve(m_count);
+
+    for (int i = 0; i < m_count; ++i)
+        weights.push_back((this->*weightGetterFuncs[i])());
+
+    return weights;
+}
+
+void MorphController::setWeight0(float weight0)
+{
+    if (weight0 == m_weights[0])
```

Comment:
Use a fuzzy comparison here

Reply:
I don't think we want to do that, we want to compare the exact values rather than whether the two are close values.